### PR TITLE
MINIFICPP-60: Add initial implementation of Site to Site changes.

### PR DESCRIPTION
**Pull request by phrocker**
MINIFICPP-60: Update interfaces to allow for seamless extension

MINIFICPP-60: Update functions to accomodate changed signatures

Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by phrocker** [comment]
I'll be going through this myself over the next few days as well, but I like to leave the comments for others to see on the PR

File: extensions/http-curl/client/HTTPClient.cpp [hunk 1]
```diff
@@ -175,19 +201,53 @@ bool HTTPClient::submit() {
...
-
+  if (callback == nullptr) {
+    content_.ptr = &read_callback_;
+    curl_easy_setopt(http_session_, CURLOPT_WRITEFUNCTION, &utils::HTTPRequestResponse::recieve_write);
+    curl_easy_setopt(http_session_, CURLOPT_WRITEDATA, static_cast<void*>(&content_));
+  }
   curl_easy_setopt(http_session_, CURLOPT_HEADERFUNCTION, &utils::HTTPHeaderResponse::receive_headers);
   curl_easy_setopt(http_session_, CURLOPT_HEADERDATA, static_cast<void*>(&header_response_));
-
+  curl_easy_setopt(http_session_, CURLOPT_TCP_KEEPALIVE, 0L);
```
Comment (phrocker) [inline extensions/http-curl/client/HTTPClient.cpp]: This shouldn't be necessary.

File: libminifi/test/curl-tests/sitetositehttp/HTTPHandlers.h [hunk 2]
```diff
@@ -0,0 +1,322 @@
...
+
+  explicit DeleteTransactionResponder(const std::string base_url, std::string response_code, moodycamel::ConcurrentQueue<std::shared_ptr<FlowObj>> *feed)
+      : base_url(base_url),
+        response_code(response_code),
+        flow_files_feed_(feed) {
+  }
+
+  bool handleDelete(CivetServer *server, struct mg_connection *conn) {
+
+    std::string site2site_rest_resp = "";
```
Comment (phrocker) [inline libminifi/test/curl-tests/sitetositehttp/HTTPHandlers.h]: Should do some additional verification here if possible.

**Comment by achristianson** [comment]
Nothing glaring on initial glance. There are a lot of changes. Can you summarize generally what the changes are for? Just want to get the right understanding to put it into context.

File: extensions/http-curl/client/HTTPCallback.h [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+    std::vector<char> vec;
+
+    if (size > 0) {
+      vec.resize(size);
+
+      memcpy(vec.data(), vector, size);
+
+      size_t added_size = vec.size();
+
+      byte_arrays_.enqueue(std::move(vec));
```
Comment (minifirocks) [inline extensions/http-curl/client/HTTPCallback.h]: do we need to lock the mutex_ before enqueue?

File: extensions/http-curl/client/HTTPCallback.h [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+    return rolling_count_;
+  }
+
+ private:
+
+  inline void load_buffer() {
+    std::unique_lock<std::recursive_mutex> lock(mutex_);
+    cv.wait(lock, [&] {return byte_arrays_.size_approx() > 0 || is_alive_==false;});
+    if (!is_alive_ && byte_arrays_.size_approx() == 0) {
+      return;
```
Comment (minifirocks) [inline extensions/http-curl/client/HTTPCallback.h]: unlock the mutex before return?

Comment (phrocker) [reply extensions/http-curl/client/HTTPCallback.h]: wait releases the mutex

File: extensions/http-curl/client/HTTPCallback.h (hunk 3 above)
Comment (phrocker) [reply extensions/http-curl/client/HTTPCallback.h]: byte_arrays_ is a lock free queue. There can be concerns with using lock free queues like moodycamel, particularly with ordering, but I've mitigated this by how I'm using it.

File: extensions/http-curl/client/HTTPCallback.h (hunk 4 above)
Comment (minifirocks) [reply extensions/http-curl/client/HTTPCallback.h]: i thought after wait fired, we own the lock, that's reason you did unlock afterwards.

Comment (phrocker) [reply extensions/http-curl/client/HTTPCallback.h]: Sorry I thought you meant in the wait, assuming it can't get the lock. My apologies I misunderstood.

I didn't put a release there since every other usage of the lock would free and we wouldn't continue operations if we have no data or are not alive; however, I should be consistent.

File: extensions/http-curl/sitetosite/HTTPProtocol.cpp [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+}
+
+int HttpSiteToSiteClient::readRespond(const std::shared_ptr<Transaction> &transaction, RespondCode &code, std::string &message) {
+  if (current_code == FINISH_TRANSACTION) {
+
+    if (transaction->getDirection() == SEND) {
+      auto stream = dynamic_cast<io::HttpStream*>(peer_->getStream());
+      stream->closeStream();
+      auto client = stream->getClient();
+      logger_->log_info("Faking it till I makin it 2, response code %d", client->getResponseCode());
```
Comment (minifirocks) [inline extensions/http-curl/sitetosite/HTTPProtocol.cpp]: the log info print is not clear

File: extensions/http-curl/sitetosite/HTTPProtocol.cpp [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+      if (client->getResponseCode() == 202) {
+        code = CONFIRM_TRANSACTION;
+        message = std::string(client->getResponseBody().data(), client->getResponseBody().size());
+      } else {
+        code = UNRECOGNIZED_RESPONSE_CODE;
+      }
+      return 1;
```
Comment (minifirocks) [inline extensions/http-curl/sitetosite/HTTPProtocol.cpp]: return -1 for failure

File: extensions/http-curl/sitetosite/HTTPProtocol.cpp [hunk 7]
```diff
@@ -0,0 +1,312 @@
...
+    return 1;
+  }
+  return SiteToSiteClient::readRespond(transaction, code, message);
+
+}
+// write respond
+int HttpSiteToSiteClient::writeRespond(const std::shared_ptr<Transaction> &transaction, RespondCode code, std::string message) {
+  current_code = code;
+  if (code == CONFIRM_TRANSACTION || code == FINISH_TRANSACTION) {
+    logger_->log_info("Faking it till I makin it");
```
Comment (minifirocks) [inline extensions/http-curl/sitetosite/HTTPProtocol.cpp]: log print is not clear

File: extensions/http-curl/sitetosite/HTTPProtocol.cpp (hunk 6 above)
Comment (phrocker) [reply extensions/http-curl/sitetosite/HTTPProtocol.cpp]: If we return -1 we won't have an opportunity to provide an error message. -1 would be for the case where there is a comms error. here we have no need to read and write for respond. But you did remind me to change that name to readResponse.

File: extensions/http-curl/client/HTTPClient.cpp [hunk 8]
```diff
@@ -204,7 +264,9 @@ const char *HTTPClient::getContentType() {
 }
 
 const std::vector<char> &HTTPClient::getResponseBody() {
```
Comment (achristianson) [inline extensions/http-curl/client/HTTPClient.cpp]: c++11+ guarantees std::string bytes are sequential. May not need a vector<char> in lieu of string.

Comment (achristianson) [reply extensions/http-curl/client/HTTPClient.cpp]: Not in these changes, but wanted to note...

File: extensions/http-curl/processors/InvokeHTTP.cpp [hunk 9]
```diff
@@ -33,6 +33,7 @@
 #include <utility>
 #include <vector>
 
+#include "../../../libminifi/include/utils/ByteArrayCallback.h"
```
Comment (achristianson) [inline extensions/http-curl/processors/InvokeHTTP.cpp]: Awkward include path. Perhaps include path could be set better in CMakeLists.txt?

File: extensions/http-curl/processors/InvokeHTTP.h [hunk 10]
```diff
@@ -23,14 +23,14 @@
 #include <regex>
 
 #include <curl/curl.h>
+#include "../../../libminifi/include/utils/ByteArrayCallback.h"
```
Comment (achristianson) [inline extensions/http-curl/processors/InvokeHTTP.h]: Ditto.

File: extensions/http-curl/processors/InvokeHTTP.cpp (hunk 9 above)
Comment (phrocker) [reply extensions/http-curl/processors/InvokeHTTP.cpp]: cmake lists? huh?

no, that's eclipse. it does that to me. thanks!

File: extensions/http-curl/protocols/RESTSender.h [hunk 11]
```diff
@@ -22,12 +22,13 @@
 #include "json/writer.h"
 #include <string>
 #include <mutex>
+
+#include "../../../libminifi/include/utils/ByteArrayCallback.h"
```
Comment (achristianson) [inline extensions/http-curl/protocols/RESTSender.h]: Ditto.

File: extensions/http-curl/client/HTTPClient.cpp (hunk 8 above)
Comment (phrocker) [reply extensions/http-curl/client/HTTPClient.cpp]: yes and it changes again in 17...some not for the best unfortunately.

herb sutter has some really good talks about small string optimization and how you shouldn't fight it, but I'm not too certain we can really avoid the pessimization.  what are your thoughts on the SOO causing pessimization against strings that are larger than the buffer on things like sizeof, etc? I've seen profiles where it can really cause unnecessary havoc.

File: extensions/http-curl/protocols/RESTSender.h (hunk 11 above)
Comment (phrocker) [reply extensions/http-curl/protocols/RESTSender.h]: oh man sometimes eclipse really does this to me. thanks for pointing these out. I should really disable that feature

File: extensions/http-curl/sitetosite/HTTPProtocol.cpp [hunk 12]
```diff
@@ -0,0 +1,310 @@
+/**
```
Comment (achristianson) [inline extensions/http-curl/sitetosite/HTTPProtocol.cpp]: General organizational note --it seems to be a bit of a mixing of concerns to have s2s-specific code under the http-curl extension.

Granted, we probably have/will have dependencies between extensions which may or may not be easy to express using cmake constructs.

Comment (phrocker) [reply extensions/http-curl/sitetosite/HTTPProtocol.cpp]: this HTTPProtocol implementation uses CURL to do its work, hence why it exists here. While we could certainly have cross compilation between extensions, it doesn't seem like it's worth the increase in build time to change. One thing I want to avoid is having single function extensions and instead of have classes of extensions.

File: extensions/http-curl/sitetosite/HTTPProtocol.h [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+        logger_(logging::LoggerFactory<HttpSiteToSiteClient>::getLogger()) {
+    peer_ = std::move(peer);
+    peer_state_ = READY;
+  }
+  // Destructor
+  virtual ~HttpSiteToSiteClient() {
+
+  }
+
+  void setPeer(std::unique_ptr<SiteToSitePeer> peer) {
```
Comment (achristianson) [inline extensions/http-curl/sitetosite/HTTPProtocol.h]: setPeer is not virtual.

File: extensions/http-curl/sitetosite/HTTPProtocol.h [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+    peer_ = std::move(peer);
+  }
+
+  virtual bool getPeerList(std::vector<PeerStatus> &peers);
```
Comment (achristianson) [inline extensions/http-curl/sitetosite/HTTPProtocol.h]: New c++ idiom is to mark this override rather than virtual. Helps better express intent and catch bugs at compile time.

Comment (phrocker) [reply extensions/http-curl/sitetosite/HTTPProtocol.h]: Good point. I don't want to yet do that in core but the intent was to do that in extensions. Thanks for bringing that up.

File: extensions/http-curl/sitetosite/HTTPProtocol.h [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+    uri.append(std::to_string(peer_->getPort()));
+    uri.append("/nifi-api/");
+    return uri;
+  }
+
+  virtual void tearDown();
+
+  const std::string parseTransactionId(const std::string &uri);
+
+  std::unique_ptr<utils::HTTPClient> create_http_client(const std::string &uri, const std::string method = "POST", bool setPropertyHeaders = false) {
```
Comment (achristianson) [inline extensions/http-curl/sitetosite/HTTPProtocol.h]: Make method a reference?

**Comment by achristianson** [comment]
+1

**Comment by minifirocks** [comment]
@phrocker please resolve the conflict and merge into simple commit, update readme for the same.
